learning-another-language/js-loops.md

```diff
@@ -0,0 +1,417 @@
+	console.log(arr[i]);
+}
+
+// Output:
+// foo
+// bar
+// baz
+```
+
+## 2.`for … of` loop
+```js
+for (variable of object) {
+	// do something
+}
+```
+
+The `for … of` loop can be used to iterate over arrays, sets, and maps. When using the `for … of` loop, there’s no danger of going out of bounds! The loop will go over each item in the iterable object.
+
+### Examples
+
+**1. Iterate over items in a list**
+```js
+const monsters = [
+	"Michael Myers", 
+	"Jason Voorhees", 
+	"Freddy Krueger"
+];
+
+for (let monster of monsters) {
+	console.log(monster);
+}
+
+// Output:
+// Michael Myers
+// Jason Voorhees
+// Freddy Krueger
+```
+
+**2. Iterate over items in a map**
+```js
+const m = new Map();
+m.set(1, "red");
+m.set(2, "black")
+
+for (let item of m) {
+	console.log(item);
+}
+
+// Output:
+// [1, red]
+// [2, black]
+```
+
+## 3. `for … in` loop
+```js
+for (property in object) {
+	// do stuff
+}
+```
+
+The `for … in` loop iterates over all enumerable properties of an object. The code in the loop is executed for each property in the object. 
+
+### Example
+```js
+const movies = {
+	2008: "The Dark Knight",
+	2009: "Avatar",
+	2010: "Toy Story 3"
+};
+
+for (let key in movies) {
+	console.log(key + ": " + movies[key]);
+}
+
+// Ouput:
+// 2008: The Dark Knight
+// 2009: Avatar
+// 2010: Toy Story 3
+```
+
+### Common Pitfall: Unexpected Behavior When Iterating over an Array
+
+While you can use the `for … in` to iterate over an array, it is recommended to use one of the aforementioned for-loops instead. 
+
+The reason for this is because the `for … in` loop iterates over **all** of the enumerable properties in an object, including any that may be inherited.
+
+If, for instance, a JS library modifies the `Array` prototype, the loop will iterate over any additional properties as well. Check out the example below for a potentially non-intuitive demonstration.
+
+```js
+const arr = [1, 2, 3];
+
+Array.prototype.newMethod = true;
+
+for (const i in arr) {
+	console.log(i);
+}
+
+// Output:
+// 1
+// 2
+// 3
+// newMethod
+```
+
+It is also worth mentioning that the `for … in` loop is meant for objects, and thus will be slower than other loops which are better suited for arrays.
+
+### Practical Usage
+
+So, wait, why use the `for … in` loop at all?
+
+The `for … in` loop may be most practically used as a tool to debug. It is an easy way to check the properties of an object. 
+
+## 4. while loop
+```js
+while (condition) {
+	// do some repetitive task
+}
+```
+
+The `while` loop will evaluate the `condition` before the loop is run each time. If the `condition` evaluates to `true`, the loop executes the statement(s) in the block. Otherwise, the loop exits and stops executing the statement(s) in the block.
+
+The `while` loop is known as a pre-test loop because it evaluates the `condition` *before* each iteration. The loop will never execute if the condition evaluates to `false` before entering the loop.
```

Comment:
Consider eliminating the mention of pre-test loops. In my view this is extra jargon that they won't come across much. 

```suggestion
`While` loops evaluate the `condition` *before* each iteration. The loop will never execute if the condition evaluates to `false` before entering the loop.
```